pallets/pooled-staking/src/calls.rs

```diff
@@ -0,0 +1,522 @@
+                let shares = pools::AutoCompounding::<T>::shares(&candidate, &delegator);
+                let stake = pools::AutoCompounding::<T>::shares_to_stake(&candidate, shares)?;
+                pools::AutoCompounding::<T>::set_hold(&candidate, &delegator, stake);
+                (held, stake)
+            }
+            AllTargetPool::ManualRewards => {
+                let held = pools::ManualRewards::<T>::hold(&candidate, &delegator);
+                let shares = pools::ManualRewards::<T>::shares(&candidate, &delegator);
+                let stake = pools::ManualRewards::<T>::shares_to_stake(&candidate, shares)?;
+                pools::ManualRewards::<T>::set_hold(&candidate, &delegator, stake);
+                (held, stake)
+            }
+            AllTargetPool::Leaving => {
+                let held = pools::Leaving::<T>::hold(&candidate, &delegator);
+                let shares = pools::Leaving::<T>::shares(&candidate, &delegator);
+                let stake = pools::Leaving::<T>::shares_to_stake(&candidate, shares)?;
+                pools::Leaving::<T>::set_hold(&candidate, &delegator, stake);
+                (held, stake)
+            }
+        };
+
+        if stake == held {
+            return Ok(().into());
+        }
+
+        // Transfer is done using withdraw to ensure it works regardless of ED.
+        if let Some(diff) = stake.0.checked_sub(&held.0) {
+            T::Currency::transfer(
+                &T::StakingAccount::get(),
+                &delegator,
+                diff,
+                Preservation::Preserve,
+            )?;
+            T::Currency::hold(&T::CurrencyHoldReason::get(), &delegator, diff)?;
+            return Ok(().into());
+        }
+
+        if let Some(diff) = held.0.checked_sub(&stake.0) {
+            T::Currency::release(
+                &T::CurrencyHoldReason::get(),
+                &delegator,
+                diff,
+                Precision::Exact,
+            )?;
+            T::Currency::transfer(
+                &delegator,
+                &T::StakingAccount::get(),
+                diff,
+                Preservation::Preserve,
+            )?;
+            return Ok(().into());
+        }
+
+        // should be unreachable as diff must either be positive or negative
+        Ok(().into())
+    }
+
+    pub fn request_delegate(
+        candidate: Candidate<T>,
+        delegator: Delegator<T>,
+        pool: TargetPool,
+        stake: T::Balance,
+    ) -> DispatchResultWithPostInfo {
+        ensure!(!stake.is_zero(), Error::<T>::StakeMustBeNonZero);
+
+        // Convert stake into joining shares quantity.
+        let shares = pools::Joining::<T>::stake_to_shares_or_init(&candidate, Stake(stake))?;
+
+        // If the amount was stake and is less than the value of 1 share it will round down to
+        // 0 share. We avoid doing any work for 0 shares.
+        ensure!(!shares.0.is_zero(), Error::<T>::StakeMustBeNonZero);
```

Comment:
is the error returned correct? Should it include `Share` instead of `Stake`?

Reply:
`stake = shares * share_price` so it's not wrong, but I agree that it is a bit unintuitive that as a user you can pass a non-zero stake value and see this error.